src/Hazelcast.Net/CP/CPSubsystem.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+// http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Threading.Tasks;
+using Hazelcast.Clustering;
+using Hazelcast.Core;
+using Hazelcast.Exceptions;
+using Hazelcast.Protocol.Codecs;
+
+namespace Hazelcast.CP
+{
+    /// <summary>
+    /// Provides the <see cref="ICPSubsystem"/> implementation.
+    /// </summary>
+    internal class CPSubsystem : ICPSubsystem
+    {
+        private readonly Cluster _cluster;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CPSubsystem"/> class.
+        /// </summary>
+        /// <param name="cluster">The cluster.</param>
+        public CPSubsystem(Cluster cluster)
+        {
+            _cluster = cluster;
+        }
+
+        // NOTES
+        //
+        // Java CP objects are managed by CPSubsystemImpl and created through ClientRaftProxyFactory
+        // which is a simplified factory, which does not cache AtomicLong, AtomicRef, and CountDownLatch,
+        // and sort-of caches (?) FencedLock and Semaphore.
+        //
+        // These objects are therefore IDistributedObject but *not* DistributedObjectBase, and *not*
+        // managed by the DistributedObjectFactory.
+        //
+        // The are destroyed via ClientProxy.destroy, which is getContext().getProxyManager().destroyProxy(this),
+        // which means they are destroyed by ProxyManager aka DistributedObjectFactory, which would try to
+        // remove them from cache (always missing) and end up doing proxy.destroyLocally() which eventually
+        // calls into the object's onDestroy() method.
+        //
+        // But... this is convoluted? For now, our objects inherit from CPObjectBase which is simpler than
```

Comment:
+1
We also have separate base classes for the Python and Node.js clients to avoid this problem